src/Reflex/PubSub.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE ExplicitForAll        #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE RecursiveDo           #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Reflex.PubSub where
+
+import Prelude                hiding (lookup)
+
+import Control.Monad.Reader
+import Data.Constraint.Extras
+import Data.Dependent.Map
+import Data.Dependent.Sum
+import Data.Functor.Identity
+
+import Reflex
+
+type PubSubT t tag m = EventWriterT t [DSum tag Identity] (ReaderT (Event t (DMap tag Identity)) m)
+
+-- | 'PubSub' enables you to broadcast events from anywhere in your reflex network to
+-- any location that is subscribed to the topic they are published to. Topics are best
+-- representated as constructors of some custom GADT which carries a phantom type to
+-- constrain the messages that can be published to it.
+class (Reflex t, Monad m, Semigroup a) => PubSub t tag m a where
+  publish :: tag a -> Event t a -> m ()
```

Comment:
Maybe add `publishes :: tag a -> (b -> a) -> Event t b -> m ()`? I'm less confident about this than I am `subscribes`, below, but it could be useful with lens: if you have `a ~ Endo c`, then you can use this to roll up a bunch of updates (generated from `Endo $ set someLens someVal`, for instance) and ship them around to different parts of your program? (Though maybe that's an anti-pattern, and the correct thing to do in that instance would be to stick a `Dynamic t` in a `ReaderT` instead of reconstructing it at each leaf?)